src/mlpack/methods/decision_tree/decision_tree.hpp

```diff
@@ -354,6 +359,39 @@ class DecisionTree :
    */
   size_t NumClasses() const;
 
+  /**
+   * Prune the already built decision tree by making tree nodes as leaf nodes
+   * if the score on the validation set increases.
+   *
+   * @param root pointer to the root node of the tree for score calculation.
+   * @param labels Labels for each training point.
+   * @param numClasses Number of classes in the dataset.
+   * @param weights Weights of all the labels
+   * @param validData Validation dataset.
+   * @param validLabels labels for validation points.
+   * @param bestScore best score of pruning on validation set.
+   */
+  template<bool UseWeights, typename MatType, typename LabelsType,
+      typename WeightsType>
+  void Prune(DecisionTree* root,
+             LabelsType&& labels,
+             const size_t numClasses,
+             WeightsType&& weights,
+             MatType&& validData,
+             LabelsType&& validLabels,
+             double& bestScore);
```

Comment:
I'd also suggest that we find a way to templatize out the strategy for pruning, instead of including code in the DecisionTree itself that specifies a single pruning strategy.

Reply:
Ya i agree with templatizing pruning() (and/or validascore() as well to include different validating strategies) . Infact that was my next step once the code was reviewed to be fine.

Reply:
We wanted to first check whether the implementation is correct. If so, we would make template for the tree. So, can you please review the implementation and the tests, so that we can move to the next step.